modyn/evaluator/internal/pytorch_evaluator.py

```diff
@@ -0,0 +1,177 @@
+        self._evaluation_transformer = self._setup_transformer_function(
+            evaluation_info.evaluation_transformer, EVALUATION_TRANSFORMER_FUNC_NAME
+        )
+
+        self._metrics = evaluation_info.metrics
+
+        self._label_tranformer_function = self._setup_transformer_function(
+            evaluation_info.label_transformer, LABEL_TRANSFORMER_FUNC_NAME
+        )
+
+        self._device = evaluation_info.device
+        self._device_type = "cuda" if "cuda" in self._device else "cpu"
+        self._amp = evaluation_info.amp
+
+        self._status_query_queue = status_query_queue
+        self._status_response_queue = status_response_queue
+
+        self._num_samples = 0
+
+        self._info("Initialized PyTorch evaluator.")
+
+    def _prepare_dataloader(self, evaluation_info: EvaluationInfo) -> torch.utils.data.DataLoader:
+        self._debug("Creating EvaluationDataset.")
+        dataset = EvaluationDataset(
+            evaluation_info.dataset_id,
+            evaluation_info.bytes_parser,
+            evaluation_info.transform_list,
+            evaluation_info.storage_address,
+            evaluation_info.evaluation_id,
+        )
+        self._debug("Creating DataLoader.")
+        dataloader = torch.utils.data.DataLoader(
+            dataset, batch_size=evaluation_info.batch_size, num_workers=evaluation_info.num_dataloaders
+        )
+
+        return dataloader
+
+    def _setup_transformer_function(self, serialized_code: str, func_name: str) -> Optional[Callable]:
+        if serialized_code == "":
+            return None
+        exec(serialized_code, self._mod_dict)  # pylint: disable=exec-used
+        if func_name not in self._mod_dict or not isfunction(self._mod_dict[func_name]):
+            raise ValueError(f"Invalid function is provided. Expected a function with name {func_name}.")
+        return self._mod_dict[func_name]
+
+    def _info(self, msg: str) -> None:
+        self.logger.info(f"[Evaluation {self._evaluation_id}] {msg}")
+
+    def _debug(self, msg: str) -> None:
+        self.logger.debug(f"[Evaluation {self._evaluation_id}] {msg}")
+
+    def _load_state(self, path: pathlib.Path) -> None:
+        assert path.exists(), "Cannot load state from non-existing file"
+
+        self._info(f"Loading model state from {path}")
+        with open(path, "rb") as state_file:
+            checkpoint = torch.load(io.BytesIO(state_file.read()))
+
+        assert "model" in checkpoint
+        self._model.model.load_state_dict(checkpoint["model"])
+
+        # delete trained model from disk
+        os.remove(path)
+
+    def send_status_to_server(self, batch_number: int) -> None:
+        self._status_response_queue.put({"num_batches": batch_number, "num_samples": self._num_samples})
+
+    def evaluate(self) -> None:
+        self._info(f"Process {os.getpid()} starts evaluation.")
+
+        self._model.model.eval()
+        with torch.no_grad():
+            batch_number = -1
```

Comment:
why is this required?

Reply:
Are you referring to `batch_number = -1`? I used it to log the total number of evaluated batches.

Reply:
Yes, you are right. I initialized it by -1 as then we get consistent results if the dataloader does not return any batches.

Reply:
I guess off by one depends on what you deem off by one, right? If you think index 0 is the first batch (which I guess I do), then it is not off by one in the trainer, is it?

Reply:
Yes but in both cases (evaluator, trainer server) we are logging the total number of processed batches. In the former version we logged the last `batch_number` seen. Since this is 0-based numbering, we are then off by 1 batch e.g we log 4 as this was the last `batch_number` however we have seen batches 0, 1, 2, 3, 4 - so 5 in total. It is very subtle and I can revert to the old version if you want.

Reply:
No, no need to rollback (this is a minor detail anyways, as you said), I am just trying to understand (but this is not blocking for this PR - feel free to answer me on element and then resolve this). I still don't see how initializing this to -1 helps here and where exactly you are referring to with logging. During the first batch, `batch_number` will immediately be 0. So I don't see how it makes a difference, unless we do +1 somewhere to get the number of processed batches, but even then, I don't see how off by one is happening and how this would help, as we will always put batch_number = 0 in the queue, as we only write to the queue within the loop. But I think I am just missing something

Reply:
Ah, I forgot to answer. Yes you are right, this was previously not documented.
Now, it is done via configuration option comment in F1-score :)